### ISSUE TITLE
Improve dashboard selector `getParameters` to avoid unnecessary heavy operations

### ISSUE THREAD

**Issue by uladzimirdev** [metabase/metabase#38502]
**Context**
Improve dashboard selector `getParameters` to avoid unnecessary heavy operations

[suggestion](https://metaboat.slack.com/archives/C04CYTEL9N2/p1707156521548289?thread_ts=1706701672.407039&cid=C04CYTEL9N2)
```
why typing is so slow is because our getParameters[ selector](https://github.com/metabase/metabase/blob/ea05f6ff8dc09692761514423400ccef0cab5af8/frontend/src/metabase/dashboard/selectors.ts#L305) is using getDashboardComplete even tho it only cares about dashboard.parameters and dashcards’ parameter_mappings

We might not need debouncing heading/text change events if we could optimize this
Basically we only need to rerun getParameters when:
- metadata changes (can be frequent when initially loading a busy dashboard, mostly stale later)
- dashboard parameters (their values are decoupled, so dashboard.parameters should only change in edit mode if I’m not missing something)
- dashcard parameter mappings (again, should only be changing while in edit mode)
```

**Comment by uladzimirdev** [comment]
tested it https://github.com/metabase/metabase/pull/38765 but didn't notice any significant change in perf
